### ISSUE TITLE
The advanced drawing characters confilct with the border (just looks weird)

### ISSUE THREAD

**Issue by dirk1763** [dan200/ComputerCraft#51]
The characters 128 to 159 split each character spot in a 2*3 grid that can be switched off and on individually. Except the bottom right pixel, but that can be archived by inverting the colors. But at the border of the computerc screen the background is bigger than the text area. So the inverted character looks strange.
![border](https://cloud.githubusercontent.com/assets/8234006/11768414/8e3837b6-a1cc-11e5-91e1-047d0b716865.PNG)

test script: http://pastebin.com/xT3U7PSp

**Comment by dan200** [comment]
Not a lot I can do about this. The 2px margin between the border and the text was there from before character backgrounds could be changed, so to maintain backwards compatibility the decision was made to extend the background colour of the edge characters into the gap (otherwise, you'd just get a fixed black border, even if you'd fille the screen blue, for example). This behaviour is the same for the teletext characters as it is for any other.

**Comment by BombBloke** [comment]
It strikes me that that a fixed black border may look better than this. Worth a try?

**Comment by dan200** [comment]
It would look better for this use case, but not most others.

**Comment by KnightMiner** [comment]
Maybe make it toggleable? That will allow you to have both advantages based on what you need

**Comment by dan200** [comment]
toggleable where by whom?

**Comment by KnightMiner** [comment]
As in a function which enables/disabled the automatic border (and maybe allows setting the color as well). Something like `term.forceScreenBorder( bool force, int color )`

**Comment by dirk1763** [comment]
my real monitor has a (tiny) black border, too :smile: 
An other approach would be to strech the characters from 128 to 159 when they are printed on the border. This way there would not backwards compatibility problem.

**Comment by dirk1763** [comment]
A third way to do it, would be that only the term.clear method sets the border colors and printing characters does not.

**Comment by oddstr13** [comment]
A solution to this might be to remove the border completely, changing the terminal border texture slightly.

If not, I think I would prefer the border being solid black, rather than having it magically extend the background color. This would allow texture packs to either remove the black border, or leave it as is.

**Comment by dan200** [comment]
I thought up a nice solution to this:

Allow the background colours of the border positions to be independently set by writing to positions 0 or width+1/height+1. term.clear() would still set all of these, and term.clearLine() would set the left and right border chars of that line, so the behaviour of many current programs will be unaffected. But the control is available to those who need it

It will take a while to implement this (especially making the window API play nice with it), so might not make it into the next beta, but we'll see.

**Comment by Lion4ever** [comment]
I think i did not fully understand your solution:
If you need to write to position 0 to change the border, wouldn't all old programs loose the border?
The multishell would not be affected, but the paint program would:

![paint](https://cloud.githubusercontent.com/assets/8234006/11787238/f7510f84-a28a-11e5-80d0-966d1ccb4d6d.PNG)
and the worm program, too
![worm](https://cloud.githubusercontent.com/assets/8234006/11787250/0929b6ac-a28b-11e5-98eb-b4e4074167a3.PNG)
Even if they are not using the advanced drawing symbols at all.
What is the argument against making a border exception for the advanced drawing characters?

**Comment by dan200** [comment]
Because you don't know the intent. If you're drawing with black foreground on a white background, you might be trying to draw black text on a white background, or you might be trying to draw black text on a white background. In one case you'd want the background to spill over, in another case you wouldn't.

**Comment by Lion4ever** [comment]
i would not want the background to spill over in any case:
Case white:
- term.clear -> whole screen white
- print white drawing char with black background on border -> border stays white (because drawing char does not change border)
- print black drawing char with white background on border -> border stays white (because drawing char does not change border)
- print white letter with black background -> border goes black (like in good old day)
Case black:
- term.clear -> whole screen black
- print white drawing char with black background on border -> border stays black (because drawing char does not change border)
- print black drawing char with white background on border -> border stays black (because drawing char does not change border)
- print black letter with white background -> border goes white

**Comment by Lion4ever** [comment]
so yeah there would be rows 0 and width +1 but it would not be nessesary to rewrite anything lua sided (including user programs) to have the same functionality.

**Comment by Wojbie** [comment]
So to sum your post you are suggesting to keep current system and simply make the drawing chars not change it when they are drawn next to it?

**Comment by Lion4ever** [comment]
in which cases?

**Comment by Wojbie** [comment]
Here for example.
![Image of Yaktocat](http://i.imgur.com/wihNGZB.png)
In your example next to character 153 there would be back border instead of white.
This whole thing is kinda wierd to thing about. Personally i am not sure if keeping backwards compatibility is worth it.

**Comment by Lion4ever** [comment]
yeah, ok, it would be unexpected behavoir for many users...
you are right.

**Comment by BombBloke** [comment]
I still like the idea of forcing a black border at all times.

I created some mockups, which might make it easier to judge the effect:

![2015-12-15_23 08 35](https://cloud.githubusercontent.com/assets/12162603/11810457/640539dc-a382-11e5-8bbe-34334746c7f4.png)
![2015-12-15_23 15 00](https://cloud.githubusercontent.com/assets/12162603/11810468/6de969dc-a382-11e5-82df-c12db81f7bda.png)
![2015-12-15_23 17 54](https://cloud.githubusercontent.com/assets/12162603/11810471/736ad256-a382-11e5-9f48-4eaa7b7d213b.png)

**Comment by Wojbie** [comment]
Have to agree with Bomb Bloke here. That would be most consistent and least complicated solution.
All other ones feel gimmicky and unintuitive to me.

**Comment by apemanzilla** [comment]
I agree as well. It's the cleanest solution IMO.

**Comment by gamax92** [comment]
I also agree with a forced black border, though it might be possible to extend that as a forced color border, where background colors would still no longer bleed out on the sides but the border area would be a configurable color.

**Comment by immibis** [comment]
Why not just remove the border?

**Comment by BTCTaras** [comment]
I think that a toggleable border would be the most intuitive approach. There'd be perfect backwards compatibility, and program authors could choose which looks nicest on their UI.

If the border is removed entirely, I'd suggest changing the screen images to wrap tightly against the edge of the drawing area, eliminating that new black space. If that makes sense.

**Comment by AEnterprise** [comment]
i agree that a toggleable border (maybe with a configurable color) would be the best solution

**Comment by apemanzilla** [comment]
Toggleable borders won't work that well with the window API. For example, in multishell - if a program changes the border color when there are two tabs open, you'd be left with a border around the top tabs as well which may not match in color.

**Comment by Wojbie** [comment]
Soo like now that something is happening again in CC world anyone still interested in this discussion? 

I still think that BombBloke solution was cleanest.

**Comment by apemanzilla** [comment]
I'm still in favor of black borders.

**Comment by gamax92** [comment]
Also still in favor of BombBloke's solution, black borders seem the best solution

**Comment by SquidDev** [comment]
Seeing as #242 has been closed, wondering if people have alternative suggestions for this issue? 

One solution I've been thinking about would be to repeat the edge pixels of the teletext characters for the border. This does look rather weird for some characters though, so I'm not sure it's ideal. For instance, @Wojbie's example would become:

![image](https://user-images.githubusercontent.com/4346137/27685056-7d9306fc-5cc4-11e7-90ea-ec0893af100e.png)

**Comment by BombBloke** [comment]
That example looks perfectly fine to me, but one of my own would become:

![ccborderexample4](https://user-images.githubusercontent.com/12162603/27692054-12540872-5d29-11e7-9aeb-987b74ed6265.png)

... although I have no problems if it really can be coded to only do it for the *teletext* characters, as you say.

**Comment by Lion4ever** [comment]
I like dans approach with 0/height+1 best so far.
You could modify it to change the background of row 0 too when you print at row 1. That way you would have the same result like in older versions and you can change it back by printing at 0 when you need to.

**Comment by apemanzilla** [comment]
I'm still of the opinion that having a static black border, or just removing the border entirely, would be the best option. I don't think we should have special behavior for certain characters or certain terminal positions.

**Comment by BombBloke** [comment]
@Lion4ever: That approach would be a blow to backwards compatibility, in that scripts which casually use term.clearLine() may end up with a random mish-mash of colours for their borders. The issues you already pointed out [here](https://github.com/dan200/ComputerCraft/issues/51#issuecomment-164492199) can be fixed within the mod, but user-made scripts aren't going to retroactively adapt themselves to the change.

Given how infrequently the drawing characters are used I would consider that particular cure to be worse than the disease.

**Comment by Lion4ever** [comment]
@BombBloke No, what i meant is that term.clear and term.clearLine set the color like dan said a year ago and printing to 0 sets them manually.
When 1 and height writes to 0 and height+1 too, you would get the same behavior you had before.

**Comment by BTCTaras** [comment]
I like SquidDev's solution. Unlike any of the other proposals, there's not going to be any backwards-compatibility hiccups, and it just sounds genuinely clean and simple for the userbase.
That is, of course, if only the box drawing characters are stretched to the edge, and not characters like in bb's figure

**Comment by Lion4ever** [comment]
Well if it is only for the teletext characters, that would work too and probably be a bit more efficient.

**Comment by oddstr13** [comment]
![Monitor border](https://thoas.feralhosting.com/oddstr13/sharex/image/2017-07-17_21-55-37.png)
If anyone should care, this is how one of my physical monitors looks. It has a ~10px, solid black border between the display edge and the plastic casing. Another of my monitors has the plastic casing going all the way to the display area.

**Why make it more complicated than needed?**

Remove the edge-stretching, and do one of:
1. Leave it a solid black border (part of texture).
2. Widen the outer texture border, so that there is no black border left.
3. Resize the texture, so that it fits the drawing area, keeping the texture border width.
